### PR TITLE
add FairMQ plugin mechanism

### PR DESCRIPTION
**Pull request by dennisklein**
Plugin Service APIs I will push later. So, do not merge this PR yet.

Ref #561

### CODE REVIEW

**Comment by dennisklein** [comment]
This PR has a dependency on #597

**Comment by dennisklein** [comment]
@rbx The existing configuration event api (`FairMQEventManager`) is not suitable for multiple plugins (There is only one callback slot per config property). This needs to be rewritten with boost::signals2::connection as we have done for the state change events or similar. Anyways, I can not finish this before the workshop. So the idea is to merge it this week without the config property subscription api.

**Comment by dennisklein** [comment]
@rbx On Linux GCC 4.9.2 I get this error which I do not understand:
```
[30/89] Building CXX object fairmq/CMakeFiles/FairMQ.dir/Plugin.cxx.o
FAILED: /usr/bin/c++   -DFairMQ_EXPORTS -DMSGPACK_FOUND -DNANOMSG_FOUND -DPROTOBUF -I../fairmq -I../ -isystem /home/dklein/fair/fairsoft_dev/include -fdiagnostics-color=always -std=c++11 -O2 -g -Wshadow -fPIC -MD -MT fairmq/CMakeFiles/FairMQ.dir/Plugin.cxx.o -MF fairmq/CMakeFiles/FairMQ.dir/Plugin.cxx.o.d -o fairmq/CMakeFiles/FairMQ.dir/Plugin.cxx.o -c /home/dklein/alfabuild/FairRoot/fairmq/Plugin.cxx
/home/dklein/alfabuild/FairRoot/fairmq/Plugin.cxx: In constructor ‘fair::mq::Plugin::Plugin(std::string, fair::mq::Plugin::Version, std::string, std::string, fair::mq::PluginServices*)’:
/home/dklein/alfabuild/FairRoot/fairmq/Plugin.cxx:19:33: error: cannot convert ‘const Version {aka const fair::mq::tools::Version}’ to ‘const int’ in initialization
 , fPluginServices{pluginServices}
                                 ^
/home/dklein/alfabuild/FairRoot/fairmq/Plugin.cxx:19:33: error: uninitialized const member ‘fair::mq::tools::Version::fkMinor’
/home/dklein/alfabuild/FairRoot/fairmq/Plugin.cxx:19:33: error: uninitialized const member ‘fair::mq::tools::Version::fkPatch’
```

**Comment by dennisklein** [comment]
On Linux GCC 6.3.1 the testsuite works for me as well as the plugin from  [FairRootGroup/FairMQPlugin_example](https://github.com/FairRootGroup/FairMQPlugin_example). 

@rbx Let's see if Travis complains and what you see on your Mac.

**Comment by dennisklein** [comment]
[GCC 4 error](https://cdash.gsi.de/viewBuildError.php?buildid=156421)

**Comment by rbx** [comment]
The `--no-as-needed` linker flags just needs a `if(NOT CMAKE_SYSTEM_NAME MATCHES Darwin)` around it, since it is not available on MacOS. Otherwise everything works!

File: fairmq/Plugin.cxx [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ ********************************************************************************/
+
+#include <fairmq/Plugin.h>
+#include <FairMQLogger.h>
+
+using namespace std;
+
+fair::mq::Plugin::Plugin(const string name, const Version version, const string maintainer, const string homepage, PluginServices* pluginServices)
+: fkName{name}
+, fkVersion{version}
```
Comment (mattkretz) [inline fairmq/Plugin.cxx]: @dennisklein Change this line to ``fkVersion(version)``

**Comment by dennisklein** [comment]
@mattkretz @rbx You are awesome! Thx

**Comment by dennisklein** [comment]
@MohammadAlTurany Be aware, that we have increased minimum required Boost version from `1.41` to `1.61` in the main [CMakeLists.txt](https://github.com/dennisklein/FairRoot/blob/c268180c1456edba1a5e126d84f49a7a549e2dce/CMakeLists.txt#L247).

**Pull request merged**